07-Threads/threads.c

```diff
@@ -49,23 +49,66 @@ void systick_handler()
 	*SCB_ICSR |= SCB_ICSR_PENDSVSET;
 }
 
-void thread_start()
+void __attribute__((naked)) thread_start()
 {
 	lastTask = 0;
 
-	/* Save kernel context */
-	asm volatile("mrs ip, psr\n"
-	             "push {r4-r11, ip, lr}\n");
-
+	/* Reset APSR before context switch.
+	 * Make sure we have a _clean_ PSR for the task.
+	 */
+	asm volatile("mov r0, #0\n"
+	             "msr APSR_nzcvq, r0\n");
 	/* To bridge the variable in C and the register in ASM,
 	 * move the task's stack pointer address into r0.
 	 * http://www.ethernut.de/en/documents/arm-inline-asm.html
 	 */
 	asm volatile("mov r0, %0\n" : : "r" (tasks[lastTask].stack));
-	/* Load user task's context and jump to the task */
-	asm volatile("ldmia r0!, {r4-r11, lr}\n"
-                    "msr psp, r0\n"
-                    "bx lr\n");
+	asm volatile("msr psp, r0\n"
+	             "mov r0, #3\n"
+	             "msr control, r0\n"
+	             "isb\n");
+	/* This is how we simulate stack handling that pendsv_handler
+	 * does. Thread_create sets 17 entries in stack, and the 9
+	 * entries we pop here will be pushed back in pendsv_handler
+	 * in the same order.
+	 *
+	 *
+	 *                      pop {r4-r11, lr}
+	 *                      ldr r0, [sp]
+	 *          stack
+	 *  offset -------
+	 *        |   16  | <- Reset value of PSR
+	 *         -------
+	 *        |   15  | <- Task entry
+	 *         -------
+	 *        |   14  | <- LR for task
+	 *         -------
+	 *        |  ...  |                             register
+	 *         -------                              -------
+	 *        |   9   | <- Task argument ---->     |   r0  |
+	 * psp after pop--<                             -------
+	 *        |   8   | <- EXC_RETURN    ---->     |   lr  |
+	 *         -------                              -------
+	 *        |   7   |                            |  r11  |
+	 *         -------                              -------
+	 *        |  ...  |                            |  ...  |
+	 *         -------                              -------
+	 *        |   0   |                            |   r4  |
+	 * psp ->  -------                              -------
+	 *
+	 * Instead of "pop {r0}", use "ldr r0, [sp]" to ensure consistent
+	 * with the way how PendSV saves _old_ context.
```

Comment:
You can add some Arm references as Piko/RT did.

Reply:
Do you mean something like this below?

https://github.com/PikoRT/pikoRT/blob/master/arch/v7m-entry.S#L141
```
/* [a] https://lists.nongnu.org/archive/html/qemu-devel/2014-12/msg01575.html
 *     ADP_Stopped_ApplicationExit is used for exit(0), anything else is
 *     implemented as exit(1). */
```

Reply:
Exactly. It would be informative if these links to Arm specific resources are landed.

Reply:
Got it.

But I have an question:
Is web link appropriate to appear in comments?

It seems we have to maintain links in comments that might be not persistent in both its web address and content.
Maybe provide the document name such as `ARM® Architecture Reference Manual
ARMv8, for ARMv8-A architecture profile (ARM DDI 0487C.a)` and the chapter is better, so that people who is interesting in the detail would find the document itself.

Reply:
I agree with the way to mention chapters and sections accordingly.